File: Code Python/05. Prediction/05.3 Compressed Sensing/Introduction.py
import numpy as np
import matplotlib.pyplot as plt
import cvxpy as cvx

# Nyquist and Shannon Sampling Theorem

dt = .001
t = np.arange(0, 1, dt)

f = np.sin(2*np.pi*10*t)

smplrate = np.arange(1, 1001, 20)
samples = f[smplrate]

plt.figure(figsize=(10, 5))
plt.subplot(2, 1, 1)
plt.title('Sufficiently sampled signal')
plt.plot(f)
plt.plot(smplrate, samples, 'xr')
plt.xlim([1, 1000])
plt.legend(['Original and reconstructed signal', 'Samples'], loc='upper right')
plt.show()

smplrate = np.arange(1, 1001, 90)
samples = f[smplrate]

plt.figure(figsize=(10, 5))
plt.subplot(2, 1, 1)
plt.title('Undersampled Signal')
plt.plot(f)
plt.plot(smplrate, samples, 'xr')
plt.plot(-np.sin(2*np.pi*1.09*t))
plt.xlim([1, 1000])
plt.legend(['Original and reconstructed signal', 'Samples','Reconstructed Signal'], loc='upper right')
plt.show()

# Compressed Sensing

dt = .01
t = np.arange(0, 1, dt)

f = np.sin(2*np.pi*2*t) + np.cos(2*np.pi*5*t)

smplrate = np.random.randint(100, size=(30, 1))
samples = f[smplrate]

plt.figure(figsize=(10, 5))
plt.title('Randomly Sampled')
plt.plot(f)
plt.plot(smplrate, samples, 'xr')
plt.xlim([1, 100])
plt.legend(['Original signal', 'Samples'])
plt.show()

Orth = np.conj(np.fft.fft(np.eye(101))/101)
S = np.zeros((len(smplrate), 101))

for i in range(1, len(samples)):
    S[i, smplrate[i]] = 1

samples = samples.reshape(len(samples), 1)

plt.figure(figsize=(10, 20))
plt.subplot(2,5,1)
plt.imshow(samples, cmap='gray')
plt.axis('off')
plt.title('Sampled Data')
plt.subplot(2,5,3)
plt.imshow(np.real(Orth), cmap='gray')
plt.axis('off')
plt.title('Phi')
plt.subplot(2,5,2)
plt.imshow(S, cmap='gray')
plt.axis('off')
plt.title('R')

samples = samples.reshape(len(samples))

PSI = S @ Orth

xL1 = cvx.Variable(101, complex='True')
obj = cvx.Minimize(cvx.norm(xL1, 1))
constraints = [cvx.sum_squares(PSI@xL1-samples) <= 1]
prob = cvx.Problem(obj, constraints)
prob.solve()
x = xL1.value
print(x)

x = x.reshape(len(x), 1)
samples = samples.reshape(len(samples), 1)

plt.subplot(2, 5, 4)
plt.imshow(np.abs(x), cmap='gray')
plt.axis('off')
plt.title('Sparse L_1 x')
plt.subplot(2, 5, 5)
plt.imshow(abs(np.linalg.pinv(S.dot(Orth)).dot(samples)), cmap='gray')
plt.axis('off')
plt.title('Not sparse L_2 x')
plt.subplot(2, 5, 7)
plt.imshow(np.real(np.dot(S, Orth)), cmap='gray')
plt.axis('off')
plt.title('Psi')
plt.show()

File: Code Matlab/05. Prediction/05.3 Compressed Sensing/Introduction.m
clc, clear, close all
%% Nyquist-Shannon

dt = .001;
t = 0:dt:1;

f = sin(2*pi*10*t);

smplrate = 1:20:1001;
samples = f(smplrate);

figure('Name','Nyquist and Shannon')
subplot(2,1,1)
title('Sufficiently sampled signal')
hold on
plot(f)
plot(smplrate,samples,'xr')
xlim([1,1000])
legend('Original and reconstructed signal','Samples')

smplrate = 1:90:1001;
samples = f(smplrate);

subplot(2,1,2)
title('Undersampled Signal')
hold on
plot(f)
plot(smplrate,samples,'xr')
plot(-sin(2*pi*1.11*t))
xlim([1,1000])
legend('Original signal','Samples','Reconstructed Signal')

%% Compressed Sensing
dt = .01;
t = 0:dt:1;

f = sin(2*pi*2*t)+cos(2*pi*5*t);

smplrate = round(rand(20,1)*101);
samples = f(smplrate);

figure('Name','Randomly Sampled')
title('Randomly Sampled')
hold on
plot(f)
plot(smplrate,samples,'xr')
xlim([1,100])
legend('Original signal','Samples')

O = conj(dftmtx(101)./length(t));
S = zeros(length(samples),101);

for i = 1:length(samples)
    S(i,smplrate(i)) = 1;
end


figure('Name','Compressed Sensing')
subplot('Position',[0.05 0.5 0.025 0.4])
imagesc(samples');
axis('off')
title('Sampled Data')
subplot('Position',[0.45 0.6 0.3 0.3])
imagesc(real(O))
axis('off')
title('\Phi')
set(gca,'FontSize',15)
subplot('Position',[0.125 0.8 0.3 0.1])
imagesc(S)
axis('off')
title('R')
set(gca,'FontSize',12)

cvx_begin;
    variable xL1(101) complex; % sparse vector of coefficients 
    minimize( 10*norm(xL1,1) );
    subject to
        S*O*xL1 == samples';
cvx_end;

subplot('Position',[0.8 0.1 0.05 0.8])
imagesc(abs(xL1)), colormap('gray')
axis('off')
title('Sparse L_1 x')
subplot('Position',[0.9 0.1 0.05 0.8])
imagesc(abs(pinv(S*O)*samples'))
axis('off')
title('Not sparse L_2 x')
subplot('Position',[0.3 0.3 0.3 0.1])
imagesc(real(S*O))
axis('off')
title('\Psi')
set(gca,'FontSize',15)
